[Refactor] New play mechanism

gnabyl opened the pull request:
This is a big refactor, which re-implements song queue and a new (better) mechanism to play songs consecutively.
Names changes:
- SongQueue => QueueManager
- QueueManager = Map<string, QueueObject>
- Each QueueObject holds information about connections and player.

This new mechanism is done using onStateChange event handler so it's supposed to be a more quality code.

gnabyl committed:
Implement class for a track

gnabyl committed:
Implement queue-object class

gnabyl committed:
Change SongQueue to QueueManager

gnabyl committed:
Implement play method

This method is dedicated to play after a search command.

gnabyl committed:
Adapt handler to new play function

gnabyl committed:
Add voiceChannel to QueueObject

gnabyl committed:
Adapt skip command to new QueueObject

gnabyl committed:
Adapt join command to new mechanism

gnabyl committed:
Adapt leave command to new mechanism

gnabyl committed:
Change search return type

Also make Track() constructor public

gnabyl committed:
Fix bug with AudioResource

ytdl function doesn't set the metadata property (no idea why)

gnabyl committed:
Fix connection commands

gnabyl committed:
Fix skip command to keeps queue

bangjdev commented:
Surprisingly good doc about audio here https://discordjs.github.io/voice/modules.html
Sorry for saying that discord's doc is shit

amtcultist commented on package.json:
```diff
@@ -3,10 +3,10 @@
   "version": "1.0.0",
   "description": "A discord music bot.",
   "main": "dist/index.js",
-  "engines" : { 
```
weird spaces in package.json

amtcultist commented on src/commands/play.ts:
```diff
@@ -1,122 +1,61 @@
-}
+  queue.addTrack(track);
+  await interaction.editReply(`Track **${track.title}** added to the queue!`);
+}
```
eof

amtcultist commented on src/model/track-data.ts:
```diff
@@ -0,0 +1,12 @@
+  onStart: () => void;
+  onFinish: () => void;
+  onError: (error: Error) => void;
+}
```
eof
all properties are require ?

gnabyl replied:
Yes I guess?

amtcultist commented on src/utils/handlers.ts:
```diff
@@ -61,7 +62,7 @@ export class Handlers {
       case Commands.SKIP:
         this.lastCommand.set(interaction.guildId, interaction.user.id, {
           command: Commands.SKIP,
-          results: await executeSkip(interaction, this.queue),
```
good use of DI

amtcultist commented on src/utils/queue-object.ts:
```diff
@@ -0,0 +1,121 @@
+    this.isLocked = true;
+
+    // Take the first item from the queue
+    const nextTrack = this.songs.shift();
```
wait we remove the song from the queue to next ? 
Gonna be hard with the loop

gnabyl replied:
that's right, maybe we should have a pointer that moves, not shifting the queue :/

amtcultist reviewed (approved):
only the shift, others are pretty decent

gnabyl committed:
Fix lint format

gnabyl committed:
Fix bug in playAfterSearch

This function looks dirty and need to be refactored because it does something duplicated with executeJoin

gnabyl committed:
Set lastcommand
